pathway-analysis/02-gene_set_enrichment_analysis.Rmd

```diff
@@ -111,77 +107,143 @@ dge_results_df <- readr::read_tsv(dge_results_file)
 head(dge_results_df)
 ```
 
-We can take the same steps we did earlier to convert these Ensembl gene IDs to Entrez IDs. 
-We have to change the `column` argument to `mapIds()` (and the downstream steps).
+Since this data frame of DGE results includes gene symbols, we do not need to perform any kind of gene identifier conversion.
+We do, however, need to check for duplicate gene symbols. 
+We can accomplish this with `duplicated()`, which returns a logical vector (e.g., `TRUE` or `FALSE`).
+The function `any()` will return `TRUE` if **any** value it is given `TRUE`.
+
+```{r any_duplicated, live = TRUE}
+any(duplicated(dge_results_df$gene_symbol))
+```
+
+This will cause a problem when we go to run GSEA.
+Let's talk more about the method itself to understand _why_.
+
+## Gene Set Enrichment Analysis
+
+_Adapted from [refine.bio examples](https://github.com/AlexsLemonade/refinebio-examples/blob/33cdeff66d57f9fe8ee4fcb5156aea4ac2dce07f/03-rnaseq/pathway-analysis_rnaseq_02_gsea.Rmd)_
+
+![](diagrams/subramanian_fig1.jpg)
+
+**Figure 1. [Subramanian, Tamayo *et al.* (2005)](https://doi.org/10.1073/pnas.0506580102).**
+
+GSEA calculates a pathway-level metric, called an enrichment score (sometimes abbreviated as ES), by ranking genes by a gene-level statistic.
+This score reflects whether or not a gene set or pathway is overrepresented at the top or bottom of the gene rankings ([Subramanian, Tamayo *et al.* 2005](https://doi.org/10.1073/pnas.0506580102); [Yu](http://yulab-smu.top/clusterProfiler-book/chapter2.html#gene-set-enrichment-analysis)) 
+
+Specifically, genes are ranked from most positive to most negative based on their statistic and a running sum is calculated by starting with the most highly ranked genes and increasing the score when a gene is in the pathway and decreasing the score when a gene is not.
+In this example, the enrichment score for a pathway is the running sum's maximum deviation from zero.
+GSEA also assesses statistical significance of the scores for each pathway through permutation testing. 
+As a result, each input pathway will have a p-value associated with it that is then corrected for multiple hypothesis testing ([Subramanian, Tamayo *et al.* 2005](https://doi.org/10.1073/pnas.0506580102); [Yu](http://yulab-smu.top/clusterProfiler-book/chapter2.html#gene-set-enrichment-analysis)).
+
+The implementation of GSEA we use in here examples requires a gene list ordered by some statistic and input gene sets. 
+When you use previously computed gene-level statistics with GSEA, it is called GSEA pre-ranked.
+
+### Removing duplicates
+
+The GSEA approach requires on discriminating between genes that are in a gene set and those that are not. 
+Practically speaking, gene sets are just collections of gene identifiers!
+When the function we use for GSEA pre-ranked gets a list with duplicated gene identifiers, it can produce unexpected results.
```

Comment:
Question for my own edification: If the algorithm can handle it, why not include both duplicate ids? Say one is up and the other is neutral... would it not be reasonable to include both as a way of "averaging" the effects?

Reply:
Thinking about this a bit more... (provided that having duplicate identifiers doesn't result in errors or strange choices under the hood) I'm not sure it would have the effect of "averaging" on the enrichment score _itself_ - you can imagine a situation where the neutral instance is not in the leading edge for a positive score, i.e., the other genes in the pathway are so overrepresented at the top of the list that the score has peaked before the value associated with the neutral instance gets added to the sum. The other thing that gives me pause is the permutation testing and NES score calculation. When the method selects random gene sets of equal sizes, what happens to the duplicate identifiers? (Are we selecting those by the number of unique gene identifiers?)

If I recall correctly, the GSEA implementations on GenePattern arbitrarily select one of the values associated with an identifier, which you can make an argument for here, too. (I would not be inclined to trust a result for an individual pathway that changed much between the arbitrary choice and the more extreme choice.)

Reply:
Yeah, I wasn't expecting any kind fo real averaging... I think that if I am understanding you correctly, you are surmizing that leaving both in would be dominated by the larger score, which I think agrees with my intuition. And yes, it would be trouble for the permutation calculations to be sure.

But I also realized that the algorithm definitely can't handle duplicates in its current implementation: named vectors can't have duplicate names. So you do have to do _something_.